### PR TITLE
Fix: no-unused-vars crash from escope workaround (fixes #2042)

### LINKED ISSUE #2042: ES6 modules: TypeError: Cannot read property 'type' of undefined
Hi.

I'm using 0.17.0, and I have the following code in a file `test.js`:

```js
import EventEmitter from 'events';
```
I have the following eslint config:

```js
"env": {
  "es6": true
},
"ecmaFeatures": {
  "modules": true
}
```

When running eslint on that file I get the following error: `TypeError: Cannot read property 'type' of undefined`

Exporting or referencing that object fixes the error:

```js
import EventEmitter from 'events';

export default class Base extends EventEmitter {
  method() {
    return 'foo';
  }
}
```

Setting a property now causes a different error `TypeError: Cannot read property 'name' of undefined`:

```js
import EventEmitter from 'events';

export default class Base extends EventEmitter {
  method() {
    this.foo = 'bar';
  }
}
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9582484-es6-modules-typeerror-cannot-read-property-type-of-undefined?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by btmills**
Since this fixes an issue that crosses two rules, I'm not really sure how to test it other than manually, which I've done. Suggestions are welcome.

### CODE REVIEW

**Comment by ilyavolodin** [comment]
You wouldn't be able to test across multiple rules with eslintTester, it wasn't build for it. I guess you could write a unittest in pure Mocha, but I think it's an overkill.

**Comment by btmills** [comment]
I figured as much. Since this is only a temporary fix until escope is updated, I'm not convinced it's entirely necessary to automate this given the work that would be required.

**Comment by nzakas** [comment]
We really need a test for this - I don't really understand what it's doing or how it's helping.

**Comment by btmills** [comment]
> I don't really understand what it's doing or how it's helping.

In `no-undef`, `fixImport` is mutating the result of escope's analysis when it adds a new `Variable` to the list. `no-unused-vars` was iterating these results and crashing on the fake variable. The fix stores the fake variables that were added, then cleans them up when the rule is done.

> We really need a test for this

I was able to add a temporary one in `lib/eslint.js`.

File: lib/rules/no-undef.js [hunk 1]
```diff
@@ -110,6 +113,7 @@ module.exports = function(context) {
             variable = new escope.Variable(node.local.name, scope);
         variable.defs.push(node);
```
Comment (nzakas) [inline lib/rules/no-undef.js]: Wouldn't the better approach be to change this to:

```js
variable.defs.push({ node: node});
```

?

Comment (btmills) [inline lib/rules/no-undef.js]: It seems like that could work. But haven't we tried in the past to avoid letting rules modify the AST?

Comment (nzakas) [inline lib/rules/no-undef.js]: This isn't the AST. It's escope's representation of the data found within it. I'm already modifying it with this workaround, but it seems like my original hack didn't correctly represent the escope structure.

Comment (btmills) [inline lib/rules/no-undef.js]: With the change you just suggested and the test from this PR:

```
1) eslint verify() should not crash due to no-undef mutating escope data:
     TypeError: Cannot read property 'loc' of undefined
      at EventEmitter.module.exports.api.report (~/code/eslint/eslint/lib/eslint.js:782:28)
      at RuleContext.report (~/code/eslint/eslint/lib/rule-context.js:98:16)
      at EventEmitter.Program:exit (~/code/eslint/eslint/lib/rules/no-unused-vars.js:136:29)
      at EventEmitter.emit (events.js:129:20)
      at Controller.controller.traverse.leave (~/code/eslint/eslint/lib/eslint.js:743:25)
      at Controller.__execute (~/code/eslint/eslint/node_modules/estraverse/estraverse.js:393:31)
      at Controller.traverse (~/code/eslint/eslint/node_modules/estraverse/estraverse.js:481:28)
      at EventEmitter.module.exports.api.verify (~/code/eslint/eslint/lib/eslint.js:728:24)
      at Context.<anonymous> (~/code/eslint/eslint/tests/lib/eslint.js:2315:20)
      at callFn (~/code/eslint/eslint/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (~/code/eslint/eslint/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:387:10)
      at ~/code/eslint/eslint/node_modules/mocha/lib/runner.js:470:12
      at next (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:312:14)
      at ~/code/eslint/eslint/node_modules/mocha/lib/runner.js:322:7
      at next (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:358:17)
```

I fear this is the same kind of rabbit hole babel-eslint gets into, trying to monkey patch until dependencies get updated.

Comment (nzakas) [inline lib/rules/no-undef.js]: Fair enough, let's go with what you have.

**Pull request merged**